### PR TITLE
Improve script API for RedisAI 1.2

### PR DESCRIPTION
**Pull request by DvirDukhan**
This PR introduces
1. `AI.SCRIPTSTORE` command which will replace `AI.SCRIPTSET`. In the new command, user will have to specify entry points for the script, all with the signature `def entry_point(tensors: List[Tensor], keys: List[Str], args: List[Str])` , so the `AI.SCRIPTSTORE` command will look like `AI.SCRIPTSTORE <key> ENTRY_POINTS <n_ep> <ep_1> ... <ep_n> SOURCE <script>`
2. Modifications to the new `AI.SCRIPTEXECUTE` command which introduced a new but awkward interface. In the new modification the user will have to specify either `INPUTS` which are RedisAI tensors (either from keyspace or from DAG execution context), `KEYS` which are the keys that the function will touch during its execution (by calling `redis.Execute` command in the function, and `ARGS` which are strings representing any value the user needs. Those values should be cast later by the user in the script itself.

- [x] compile script with entry point validation
- [x] modify `AI.SCRIPTSTORE` command
- [x] modify `AI.SCRIPTEXECUTE` command
- [x] flow test
- [x] backward comptablity
- [x] RDB v3

### CODE REVIEW

File: tests/flow/tests_commands.py [hunk 1]
```diff
@@ -165,14 +165,14 @@ def test_keys_syntax(env):
     check_error_message(env, con, "KEYS scope must be provided first for AI.SCRIPTEXECUTE command",
                         "AI.SCRIPTEXECUTE script{1} bar INPUTS 2 a{1} a{1}")
 
-    # ERR KEYS section in an inner AI.SCRIPTEXEUTE command within a DAG is not allowed.
-    check_error_message(env, con, "Already encountered KEYS scope in current command",
-                        "AI.DAGEXECUTE KEYS 1 a{1} |> AI.SCRIPTEXECUTE script{1} bar KEYS 1 a{1}")
+    # # ERR KEYS section in an inner AI.SCRIPTEXEUTE command within a DAG is not allowed.
+    # check_error_message(env, con, "Already encountered KEYS scope in current command",
```
Comment (alonre24) [inline tests/flow/tests_commands.py:169]: Why disabled?

File: tests/flow/tests_commands.py [hunk 2]
```diff
@@ -165,14 +165,14 @@ def test_keys_syntax(env):
...
+    #                     "AI.DAGEXECUTE KEYS 1 a{1} |> AI.SCRIPTEXECUTE script{1} bar KEYS 1 a{1}")
 
 
 # Todo: this test should change once the script store command is implemented.
```
Comment (alonre24) [inline tests/flow/tests_commands.py:173]: remove comment...
Should add here the scriptstore errors test (like modelstore)

File: tests/flow/tests_dag.py [hunk 3]
```diff
@@ -395,13 +395,13 @@ def test_dagexecute_modelexecute_multidevice_resnet(env):
...
                         'INPUTS', 1, image_key, 'OUTPUTS', 1, temp_key1)
 
-    check_error_message(env, con, "Wrong function name given to AI.SCRIPTEXECUTE command",
-                        'AI.DAGEXECUTE', 'KEYS', 1, '{1}',
-                        '|>', 'AI.TENSORSET', image_key, 'UINT8', img.shape[1], img.shape[0], 3, 'BLOB', img.tobytes(),
-                        '|>',
-                        'AI.SCRIPTEXECUTE',  script_name, 'wrong_fn',
-                        'INPUTS', 1, image_key,
-                        'OUTPUTS', 1, temp_key1)
+    # check_error_message(env, con, "Wrong function name given to AI.SCRIPTEXECUTE command",
```
Comment (alonre24) [inline tests/flow/tests_dag.py:398]: Why disabled?

File: tests/flow/tests_dag.py [hunk 4]
```diff
@@ -553,31 +553,31 @@ def test_dagexecute_modelexecute_multidevice_resnet_ensemble_alias(env):
...
-                        'OUTPUTS', 1, temp_key2_1,
-                        '|>',
-                        'AI.SCRIPTEXECUTE', script_name_0, 'ensemble',
-                        'INPUTS', 1, temp_key2_0,
-                        'OUTPUTS', 1, temp_key1,
-                        '|>',
-                        'AI.SCRIPTEXECUTE', script_name_0, 'post_process',
-                        'INPUTS', 1, temp_key1,
-                        'OUTPUTS', 1, class_key_0)
+    # # The 'ensamble' function in script_name_0 expect to receive 2 inputs (not 1)
```
Comment (alonre24) [inline tests/flow/tests_dag.py:556]: Why disabled instead of update the error message?

File: tests/flow/tests_pytorch.py [hunk 5]
```diff
@@ -495,7 +448,7 @@ def test_pytorch_scriptexecute_errors(env):
         check_error_message(env, con, "CROSSSLOT Keys in request don't hash to the same slot", 'AI.SCRIPTEXECUTE', 'ket{1}', 'bar', 'KEYS', 1 , '{2}', 'INPUTS', 2, 'a{1}', 'b{1}', 'OUTPUTS', 1, 'c{1}')
 
         # key doesn't exist
-        check_error_message(env, con, "tensor key is empty or in a different shard", 'AI.SCRIPTEXECUTE', 'ket{1}', 'bar', 'KEYS', 1 , '{1}', 'INPUTS', 2, 'a{1}', 'b{2}', 'OUTPUTS', 1, 'c{1}')
+        check_error_message(env, con, "CROSSSLOT Keys in request don't hash to the same slot", 'AI.SCRIPTEXECUTE', 'ket{1}', 'bar', 'KEYS', 1 , '{1}', 'INPUTS', 2, 'a{1}', 'b{2}', 'OUTPUTS', 1, 'c{1}')
```
Comment (alonre24) [inline tests/flow/tests_pytorch.py:451]: Why does the error changed?

File: tests/flow/tests_pytorch.py [hunk 6]
```diff
@@ -229,29 +229,29 @@ def test_pytorch_modelinfo(env):
...
-def test_pytorch_scriptset(env):
+def test_pytorch_scriptstore(env):
     if not TEST_PT:
         env.debugPrint("skipping {} since TEST_PT=0".format(sys._getframe().f_code.co_name), force=True)
         return
 
     con = env.getConnection()
 
-    check_error(env, con, 'AI.SCRIPTSET', 'ket{1}', DEVICE, 'SOURCE', 'return 1')
+    check_error(env, con, 'AI.SCRIPTSTORE', 'ket{1}', DEVICE, 'SOURCE', 'return 1')
```
Comment (alonre24) [inline tests/flow/tests_pytorch.py:239]: Why don't we check for the error message in these 4 tests?

File: tests/flow/tests_deprecated_commands.py [hunk 7]
```diff
@@ -7,259 +7,286 @@
 '''
 
 
-def test_modelset_errors(env):
-    if not TEST_PT:
-        env.debugPrint("skipping {} since TEST_PT=0".format(sys._getframe().f_code.co_name), force=True)
-        return
+# def test_modelset_errors(env):
```
Comment (alonre24) [inline tests/flow/tests_deprecated_commands.py:10]: Why so many tests in this file are disabled?

**Review by alonre24 — COMMENTED** [review]
Looks good, much cleaner logic for executing scripts! I added many comments about missing tests (sorry...), and I saw that many tests were disabled (most of them in `tests_deprecated_commands.py`), for a reason that I couldn't understand...

File: tests/flow/tests_pytorch.py (hunk 5 above)
Comment (DvirDukhan) [reply tests/flow/tests_pytorch.py:451]: since OSS Redis uses `getkeys-api` and we mark `INPUTS` scope in script commands as a scope that contains keys. so in our case `a{1}` and `b{2}` are not mapped to the same shared.
This moved from RedisAI error to Redis error on OSS Redis usage.

File: src/redisai.c [hunk 8]
```diff
@@ -810,7 +816,115 @@ int RedisAI_ScriptSet_RedisCommand(RedisModuleCtx *ctx, RedisModuleString **argv
...
+    AC_GetString(&ac, &devicestr, NULL, 0);
+
+    RedisModuleString *tag = NULL;
+    if (AC_AdvanceIfMatch(&ac, "TAG")) {
+        AC_GetRString(&ac, &tag, 0);
+    }
+
+    if (AC_IsAtEnd(&ac)) {
+        return RedisModule_ReplyWithError(
+            ctx, "ERR Insufficient arguments, missing script SOURCE and entry points");
```
[the remark this answers is not in the document]
Comment (DvirDukhan) [reply src/redisai.c:839]: well..this specific scenario will not happen due to the wrong arity reply for `argc`< 8.
removed

File: src/backends/torch.c [hunk 9]
```diff
@@ -321,34 +324,69 @@ RAI_Script *RAI_ScriptCreateTorch(const char *devicestr, const char *scriptdef,
...
     }
 
+    // Entry points validation
+    for (size_t i = 0; i < nEntryPoints; i++) {
+        const char *entryPoint = entryPoints[i];
+        // Existence check.
+        if (!torchScript_FunctionExists(script, entryPoint)) {
+            torchDeallocContext(script);
+            char *errMsg;
+            asprintf(&errMsg, "Fuction %s does not exists in the given script.", entryPoint);
```
[the remark this answers is not in the document]
Comment (DvirDukhan) [reply src/backends/torch.c:334]: is tested

File: docs/commands.md [hunk 10]
```diff
@@ -533,7 +584,8 @@ A `TIMEOUT t` argument can be specified to cause a request to be removed from th
 ```
 AI.SCRIPTEXECUTE <key> <function> 
 KEYS n <key> [keys...] 
-[INPUTS m <input> [input ...] | [LIST_INPUTS l <input> [input ...]]*]
+[INPUTS m <input> [input ...]
+[ARGS k <arg> [arg...]]
```
[the remark this answers is not in the document]
Comment (DvirDukhan) [reply docs/commands.md:588]: I want to validate again with @MeirShpilraien

File: docs/commands.md [hunk 11]
```diff
@@ -471,8 +519,9 @@ An array with alternating entries that represent the following key-value pairs:
 !!!!The command returns a list of key-value strings, namely `DEVICE device TAG tag [SOURCE source]`.
 
 1. **DEVICE**: the script's device as a String
-1. **TAG**: the scripts's tag as a String
-1. **SOURCE**: the script's source code as a String
+2. **TAG**: the scripts's tag as a String
+3. **SOURCE**: the script's source code as a String
+* **ENTRY_POINTS** will return an array containing the script entry points
```
[the remark this answers is not in the document]
Comment (DvirDukhan) [reply docs/commands.md:524]: added

File: src/redisai.c [hunk 12]
```diff
@@ -810,7 +816,115 @@ int RedisAI_ScriptSet_RedisCommand(RedisModuleCtx *ctx, RedisModuleString **argv
  * Todo: this is temporary until we implement the new command, for testing broadcast in DMC
  */
 int RedisAI_ScriptStore_RedisCommand(RedisModuleCtx *ctx, RedisModuleString **argv, int argc) {
-    return RedisAI_ScriptSet_RedisCommand(ctx, argv, argc);
+    // AI.SCRIPTSET <key> <device> ENTRY_POINTS 1 ep1 SOURCE blob
```
[the remark this answers is not in the document]
Comment (DvirDukhan) [reply src/redisai.c:819]: comment stands for number of args in the command

File: src/execution/parsing/script_commands_parser.c [hunk 13]
```diff
@@ -4,302 +4,255 @@
...
+        ValidateKeysArgs(ctx, &argv[localArgpos - 1], argc - (localArgpos - 1), error);
+    if (keys_validation_result <= 0)
+        return REDISMODULE_ERR;
 
-    TorchScriptFunctionArgumentType *signature = RAI_ScriptRunCtxGetSignature(sctx);
-    if (!signature) {
+    if (localArgpos >= argc) {
+        RAI_SetError(error, RAI_ESCRIPTRUN, "ERR Invalid arguments provided to AI.SCRIPTEXECUTE");
+        return REDISMODULE_ERR;
+    }
```
[the remark this answers is not in the document]
Comment (DvirDukhan) [reply src/execution/parsing/script_commands_parser.c:19]: you can say that, although not 1:1
I would like to leave this check here

File: src/execution/parsing/script_commands_parser.c [hunk 14]
```diff
@@ -4,302 +4,255 @@
...
-    size_t nlists = array_len(sctx->listSizes);
-    size_t nArguments = array_len(signature);
-    size_t nInputs = array_len(inputs);
-    size_t inputsIdx = 0;
-    size_t listIdx = 0;
+    // Check validity of keys numbers.
+    localArgpos++;
+    size_t firstKeyPos = localArgpos;
+    if (firstKeyPos + nKeys > argc) {
+        RAI_SetError(error, RAI_ESCRIPTRUN,
```
[the remark this answers is not in the document]
Comment (DvirDukhan) [reply src/execution/parsing/script_commands_parser.c:30]: same as above

File: tests/flow/tests_commands.py (hunk 1 above)
Comment (DvirDukhan) [reply tests/flow/tests_commands.py:169]: because now we allow it
moreover, if a script needs keys, it will not get it from the dag context
let's iterate this, I will leave it disabled until we reach a decision

File: src/execution/parsing/script_commands_parser.c (hunk 13 above)
[the remark this answers is not in the document]
Comment (alonre24) [reply src/execution/parsing/script_commands_parser.c:19]: What is the difference?

**Review by alonre24 — COMMENTED** [review]
Added few more comments. Also, there are still some parsing errors in script execute that are not tested. I'm suppose to talk with Meir on Sunday about making the KEYS not mandatory, then we could close this PR...

File: tests/flow/test_torchscript_extensions.py [hunk 15]
```diff
@@ -36,31 +36,30 @@ def __init__(self):
...
 
     def test_redis_error(self):
-        check_error_message(self.env, self.con, "Redis command returned an error: Invalid argument",
-                            'AI.SCRIPTEXECUTE', 'redis_scripts{1}', 'test_redis_command_error',  'KEYS', 1, "x{1}",
-                            "INPUTS", 1, "x{1}",  error_msg_is_substr=True)
-        check_error_message(self.env, self.con, "Redis command returned an error: "
-                                                "WRONGTYPE Operation against a key holding the wrong kind of value",
-                            'AI.SCRIPTEXECUTE', 'redis_scripts{1}', 'test_redis_error_message',  'KEYS', 1, "hash{1}",
-                            "INPUTS", 1, "hash{1}",  error_msg_is_substr=True)
+        try:
```
Comment (alonre24) [inline tests/flow/test_torchscript_extensions.py:39]: Why this change?

**Review by alonre24 — COMMENTED** [review]
Nice! final 3 comments... and lets merge it!

**Review by alonre24 — APPROVED** [review]
